Implement discriminator feature

Catter38 opened the pull request:
Add support for discriminator properties in json to serialize/deserialize polymorphic types.
This enables usage of this library in combination with libraries from other languages like [KnownTypes](https://github.com/dmitry-bym/JsonKnownTypes) in .Net.

andreas-aeschlimann commented:
Hello, thanks for your PR.

I haven't heard of this concept before, but I have read your implementation. Is is a requirement for your backend to have the `$type` property in classes that may be extended?

Why do you want to define a list of classes that add the `$type`? Wouldn't be easier to just activate it for all classes or for none?

In any case, we would need to add some tests to that showcase your implementation. Would you mind adding some?

Catter38 commented:
Thanks for your prompt feedback! :smiley:

> I haven't heard of this concept before, but I have read your implementation. Is is a requirement for your backend to have the `$type` property in classes that may be extended?

The reason I use the `$type` property, is to easily identify which class the object is.
Of course you could just check which properties the object has and instantiate the correct type accordingly, but this would mean, that you have to change this 'creation code' every time you change the models.
With this solution, the deserializer takes care of the problem and instantiates the right type according to the `$type` property.


> Why do you want to define a list of classes that add the `$type`? Wouldn't be easier to just activate it for all classes or for none?

The reason for the class list is, that we need the class reference in the deserialization process to instantiate the class.
I couldn't find a way to create this class reference from the class name like in C# where you could use reflection to do it.
That's why I added the class list to defined the class references of the types which are subclasses.

Even if we would activate it for all classes, the problem would still be, that the type defined in the class which should be deserialized, would be the base class and the only way to know which subclass to use is the string in the `$type` property.

Do you know a way to create instances of classes by classname in typescript? :thinking:
If yes, we could remove the class list entirely.


> In any case, we would need to add some tests to that showcase your implementation. Would you mind adding some?

I've added tests now :wink:

andreas-aeschlimann commented:
Great, thanks a lot for your update! I understand now what you tried to achieve.

I was thinking if it would be possible to add classes automatically by manipulating the @JsonObject decorator function. But I think it's not possible to get the class reference inside the decorator function.

Otherwise your code seems fine. To be honest, it would maybe make sense to define the `animal` property as union type of `Cat|Dog` here instead of `Animal`. I assume one would normally like to use the child class and not the parent class properties, so it would be more useful to the developer. It is not wrong of course to write `Animal` as type.

andreas-aeschlimann commented on src/json2typescript/json-convert.ts:
```diff
@@ -669,6 +763,12 @@ export class JsonConvert {
         // Map the property
         try {
             json[jsonPropertyName] = customConverter !== null ? customConverter.serialize(classInstancePropertyValue) : this.verifyProperty(expectedJsonType, classInstancePropertyValue, true);
+
+            const classConstructorName = dataObject?.constructor?.name;
+
+            if (this._useDiscriminator && this._classes[classConstructorName]) {
+                json[this._discriminatorPropertyName] = classConstructorName;
+            }
```
I tested your changes and I found a potential issue here.

The local (TypeScript) constructor name (for example `Dog`) does not necessarily match the `$type` property coming from the server. For example, the server property `$type` could be defined as string `"DogDto"`.

Now imagine your local `classes` are defined as `classes = { "DogDto": Dog }`. In this case, your deserialization works as expected, but the serialization here fails. When doing serialization, we should not set `$type` to `Dog` (this is what the code is doing as of now), but it should be set again to `DogDto`! Basically, we need to compare the `classConstructorName` to the constructor names from all values (not keys!) of the `classes` data.

Your code works in the assumption that `$type` from the backend exactly matches the class name in the frontend. Is this a necessary convention? In any case, I think we should improve this. If you agree, I will add a commit.

andreas-aeschlimann replied:
Assuming `this.classes` is now a `Map`, then this is my suggestion for the serialization:

```
const classConstructorName = dataObject?.constructor?.name;

if (this._useDiscriminator) {
    this.classes.forEach((classDataObject: {new(): any}, key: string) => {
        if (classDataObject.name === classConstructorName) {
             json[this._discriminatorPropertyName] = key; // use the Map's key and NOT the classConstructorName!
         }
    });
}
```

Catter38 replied:
Thanks for testing the code and the great feedback!
Of course it makes a lot more sense to use the name provided in the list instead of the constructor name.
I just didn't think of this, since I always used the same string for the name and the class itself.

Your solution looks fine, so please push it :smiley:

andreas-aeschlimann commented:
Unfortunately, I cannot directly push into this pull request. But I pushed it to a seperate branch. Can you look at it and tell me your thoughts?

https://github.com/appvision-gmbh/json2typescript/commits/pullrequests/Catter38/master

I have basically done three things:
- Added ReadMe (look for discriminator), direct link: https://github.com/appvision-gmbh/json2typescript/blob/pullrequests/Catter38/master/README.md
- I implemented what we have discussed, but I simplified the "declaration" of discriminator classes. You have to call the method `registerClasses(Cat, Dog)`. It will automatically use the key defined in the `@JsonObject()` decorator. This means, you can easily register and unregister your classes on the fly without having to the deal with the id/key. In your case, it will still work if you use `@JsonObject("Cat")` for the `Cat` class.
- Adjusted and added some tests.

Catter38 commented:
I like the idea of the `registerClasses` method. It's a way cleaner way to do it.
The ReadMe also looks complete and very detailed.
I've merged your branch, so you can complete the pullrequest!

andreas-aeschlimann commented:
Cool, thanks for your input. 👍

andreas-aeschlimann commented:
Hey @Catter38, I hope you are fine. I tried to implement the discriminator feature in one of our projects and I noticed some problems. Probably this is just a documentation issue (README) and not a coding issue.

The README says you can use `jsonConvert.deserialize(animals)` and then it will automatically deserialize to Cats and Dogs. I think that is wrong: The second parameter is always required, for example `Animal`. Also, the auto-deserialization seems only to work on child properties, not on the main (JSON) object.

I haven't tested it yet, but deserialization should work fine for nested properties. I checked the tests added in this PR and the discriminator feature was used for child properties (for example `animalOwner.animal`).

Should we make it possible to add the discriminator feature on the main level as well? Otherwise the docs are not completely correct.

Catter38 commented:
Hi @andreas-aeschlimann, I'm using the library in one of our projects, and it is working fine. After inspection, I noticed, that I'm using an abstraction class to interface with json2typescript, and there I found the reason, why it's working for me. I use a dictionary in which the class references and names are stored and on deserialization, it reads the `$type` property of the root object and chooses the correct class reference.
Would be great, if we could move this part into the library itself, since it looks more like a hack right now, and of course the library should be able to deserialize the root object correctly.

But while testing, I also noticed another issue. If I try to deserialize an array of type `Animal` (as nested property) which contains `Cat` and `Dog` objects, all items will be deserialized as `Animal` and not as `Cat` or `Dog`.

Regarding the README, if we fix the deserialization of the root property, i think the README should be correct again, since it should auto detect the type by it's discriminator property.

If you want, I can take a look and try to fix the issues.

andreas-aeschlimann commented:
@Catter38 as you can see, I haven't released v1.5.0 yet, so it would be nice if we could add this before the final release.

I wasn't aware of this issue until yesterday. If you have time to do it, it would be amazing!

Catter38 commented:
@andreas-aeschlimann 
I've fixed the feature in my [fork](https://github.com/Catter38/json2typescript).
Would be great if you could take a look at it and check if you're ok with it.  :smiley:

andreas-aeschlimann commented:
Thank you @Catter38, I will hopefully be able to try it out in the next following days.
